obspy/signal/src/lanczos_resampling.c

```diff
@@ -0,0 +1,167 @@
+/*--------------------------------------------------------------------
+# Filename: lanczos_resampling.c
+#  Purpose: Lanczos resampling with different kernels
+#           each point.
+#   Author: Lion Krischer
+# Copyright (C) 2015 Lion Krischer and Martin van Driel
+#---------------------------------------------------------------------*/
+
+/* Windows apparently needs this */
+#define _USE_MATH_DEFINES
+
+#include <math.h>
+
+
+enum lanczos_window_type {
+    LANCZOS = 0,
+    HANNING = 1,
+    BLACKMAN = 2,
+};
+
+
+/* Sinc function */
+double sinc(double x) {
+    if (fabs(x) < 1E-10) {
+        return 1.0;
+    }
+    return sin(M_PI * x) / (M_PI * x);
+}
+
+
+/* Standard Lanczos Kernel */
+double lanczos_kernel(double x, int a) {
+    return sinc(x / (double)a);
+}
+
+
+/* von Hann window or raised cosine window*/
+double hanning_kernel(double x, int a) {
+    return 0.5 * (1.0 + cos(x / (double)a * M_PI));
+}
+
+
+/* Blackman window
+ *
+ * Values are taken from http://mathworld.wolfram.com/BlackmanFunction.html
+ */
+double blackman_kernel(double x, int a) {
+    return 21.0 / 50.0 +
+           0.5 * cos(x / (double)a * M_PI) +
+           2.0 / 25.0 * cos(2.0 * x / (double)a * M_PI);
+}
+
+
+/* Lanczos resampling with different kernels.
+ *
+ * Parameters:
+ *
+ *     y_in: The data values to be interpolated.
+ *     y_out: The output array. Must already be initialized with zeros.
+ *     dt: The sampling rate factor.
+ *     offset: The offset of the first sample in the output array relative
+ *         to the input array.
+ *     len_in: The length of the input array.
+ *     len_out: The length of the output array.
+ *     a: The width of the taper in samples on either side.
+ *     lanczos_window_type: Which taper window to choose.
+ *
+ * Output will be written to y_out.
+ */
+void lanczos_resample(double *y_in, double *y_out, double dt, double offset,
+                      int len_in, int len_out, int a,
+                      enum lanczos_window_type window) {
+
+    int idx, i, m;
+    double x, _x;
+
+    for (idx=0; idx < len_out; idx++) {
+        x = dt * idx + offset;
+        for (m=-a; m<=a; m++) {
+            i = (int)floor(x) - m;
+            if (i < 0 || i >= len_in) {
+                continue;
+            }
+            _x = x - i;
+            /* Apply kernel and sum up if within 'a' samples on either side. */
+            if (-a <= _x && _x <= a) {
+                if (window == LANCZOS) {
```

Comment:
It should be able to inline if they're static. But I guess compilers are just too good and loop unrolling and such these days for this change to make much difference any more.